File: web/blocks/attributes/num-attribute.dart
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

part of NetTango;

//-------------------------------------------------------------------------
/// Represents a number parameter
//-------------------------------------------------------------------------
abstract class NumAttribute extends Attribute {

  num value = null;
  num defaultValue = null;

  String getValue() {
    if (value == null) { return ""; }
    String valueString = (value != null ? value : 0).toStringAsFixed(1);
    return (valueString.endsWith(".0")) ? valueString.substring(0, valueString.length - 2) : valueString;
  }
  void setValue(String valueString) => value = NumUtils.toNum(valueString, 0.0);

  String getDefaultValue() => defaultValue == null ? "" : defaultValue.toString();
  void setDefaultValue(String valueString) => defaultValue = NumUtils.toNum(valueString, null);

  /// represents a random number?
  bool random = null;

  /// step interval for selections (for numbers and range)
  num stepSize = 1;

  // Perhaps surprisingly, this class does *not* correspond to the `"num"` attribute `type`.
  // That type is for the `ExpressionAttribute`.  This class can be `int` or `range`.
  // -Jeremy B July 2020
  NumAttribute(Block block, int id) : super(block, id);

  NumAttribute.clone(Block block, NumAttribute source, bool isSlotBlock) : super.clone(block, source, isSlotBlock) {
    this.random = source.random;
    this.stepSize = source.stepSize;
    this.defaultValue = source.defaultValue;
    if (!isSlotBlock) {
      this.value = (source.value == null) ? this.defaultValue : source.value;
    }
  }

  bool shouldQuote() { return false; }

  void showParameterDialog(int x, int y, Function acceptCallback) {
    final backdrop = block.workspace.backdrop;
    backdrop.classes.add("show");
    final dialog = block.workspace.dialog;
    dialog.style.top = "${y}px";
    dialog.classes.remove("small");
    dialog.innerHtml = "";

    String inputCode = buildHTMLInput();

    dialog.appendHtml("""
      <div class="nt-param-table">
        <div class="nt-param-row">${inputCode}</div>
      </div>
      <button class="nt-param-confirm" type="button">OK</button>
      <button class="nt-param-cancel" type="button">Cancel</button>
    """);

    HtmlElement label = querySelector("#nt-param-label-$uniqueId");
    InputElement input = querySelector("#nt-param-$uniqueId");

    querySelectorAll(".nt-param-confirm").onClick.listen((e) {
      if (input != null) {
        setValue(input.value);
      }
      backdrop.classes.remove("show");
      acceptCallback();
      final formattedValue = CodeFormatter.formatAttributeValue(this);
      block.workspace.programChanged(new AttributeChangedEvent(this.block.id, this.block.instanceId, this.id, this.type, this.value, formattedValue));
    });

    querySelectorAll(".nt-param-cancel").onClick.listen((e) => backdrop.classes.remove("show") );

    if (input != null) {
      input.focus();
      if (label != null) {
        input.onChange.listen((e) { label.innerHtml = input.value; });
        input.onInput.listen((e) { label.innerHtml = input.value; });
      }
    }
  }

  String buildHTMLInput() {
    return """
      <div class="nt-param-name">${name}</div>
      <div class="nt-param-value">
        <input class="nt-param-input" id="nt-param-${uniqueId}" type="number" step="${stepSize}" value="${value}">
        <span class="nt-param-unit">${unit}</span>
      </div>
    """;
  }
}

File: src/blocks/attributes/num-attribute.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import { IntAttributeInput, NumAttributeInput, NumberValueInput, RangeAttributeInput } from "../../types/types"
import { NumUtils } from "../../utils/num-utils"
import { Block } from "../block-instance"
import { CodeFormatter } from "../code-formatter"
import { AttributeChangedEvent } from "../program-changed-event"
import { Attribute } from "./attribute"

//-------------------------------------------------------------------------
/// Represents a number parameter
//-------------------------------------------------------------------------
abstract class NumAttribute extends Attribute {

  readonly numDef: NumAttributeInput
  readonly na: NumberValueInput

  setValue(valueString: string) {
    this.na.value = NumUtils.toNum(valueString, this.numDef.default)
  }

  // Perhaps surprisingly, this class does *not* correspond to the `"num"` attribute `type`.
  // That type is for the `ExpressionAttribute`.  This class can be `int` or `range`.
  // -Jeremy B July 2020
  constructor(numDef: IntAttributeInput | RangeAttributeInput, na: NumberValueInput, block: Block) {
    super(numDef, na, block)
    this.numDef = numDef
    this.na = na
  }

  showParameterDialog(x: number, y: number, acceptCallback: () => void): void {
    const backdrop = this.block.workspace.backdrop
    backdrop.classList.add("show")
    const dialog = this.block.workspace.dialog
    dialog.style.top = `${y}px`
    dialog.classList.remove("small")
    dialog.innerHTML = ""

    const inputCode = this.buildHTMLInput()

    dialog.insertAdjacentHTML("beforeend", `
      <div class="nt-param-table">
        <div class="nt-param-row">${inputCode}</div>
      </div>
      <button class="nt-param-confirm" type="button">OK</button>
      <button class="nt-param-cancel" type="button">Cancel</button>
    `)

    const label = document.querySelector(`#nt-param-label-${this.uniqueId}`) as Element
    const input = document.querySelector(`#nt-param-${this.uniqueId}`) as HTMLInputElement

    document.querySelectorAll(".nt-param-confirm").forEach( (el) =>
      el.addEventListener("click", (e) => {
        if (input !== null) {
          this.setValue(input.value)
        }
        backdrop.classList.remove("show")
        acceptCallback()
        const formattedValue = NumAttribute.numberValue(this.numDef, this.na)
        if (this.block.b.instanceId === null) {
          throw new Error("Cannot show parameter dialog for a non-instance block.")
        }
        this.block.workspace.programChanged(new AttributeChangedEvent(this.block.def.id, this.block.b.instanceId, this.def.id, this.a.type, this.na.value, formattedValue))
      })
    )

    document.querySelectorAll(".nt-param-cancel").forEach( (el) =>
      el.addEventListener("click", (e) =>
        backdrop.classList.remove("show")
      )
    )

    if (input !== null) {
      input.focus()
      if (label !== null) {
        input.addEventListener("change", (e) => { label.innerHTML = input.value; })
        input.addEventListener("input", (e) => { label.innerHTML = input.value; })
      }
    }
  }

  buildHTMLInput(): string {
    return `
      <div class="nt-param-name">${this.def.name}</div>
      <div class="nt-param-value">
        <input class="nt-param-input" id="nt-param-${this.uniqueId}" type="number" step="${this.numDef.step}" value="${this.na.value}">
        <span class="nt-param-unit">${this.def.unit}</span>
      </div>
    `
  }

  getDisplayValue(): string { return `${NumAttribute.numberValue(this.numDef, this.na)}${this.def.unit}` }

  static numberValue(def: NumAttributeInput, na: NumberValueInput): string {
    const valueString: string = na.value.toFixed(NumAttribute.stepSizePrecision(def)).toString()
    return (valueString.endsWith(".0")) ? valueString.substring(0, valueString.length - 2) : valueString
  }

  static stepSizePrecision(def: NumAttributeInput): number {
    if (Number.isInteger(def.step)) {
      return 0
    } else {
      const text = def.step.toString()
      return text.length - text.indexOf('.') - 1
    }
  }

}

export { NumAttribute }
